vllm/v1/worker/utils.py

```diff
@@ -134,27 +134,33 @@ def reset_cache(self) -> None:
 @dataclass
 class AttentionGroup:
     backend: type[AttentionBackend]
+    layer_names: list[str]
+    kv_cache_spec: KVCacheSpec
+    kv_cache_group_id: int
     # When ubatching is enabled we will have a metadata builder for each ubatch
     # so that if they use internal persistant buffers for cudagraphs, and they
     # won't have to worry about conflicting with the other ubatches.
-    metadata_builders: list[AttentionMetadataBuilder]
-    layer_names: list[str]
-    kv_cache_spec: KVCacheSpec
+    metadata_builders: list[AttentionMetadataBuilder] = field(
+        default_factory=lambda: []
+    )
 
-    @staticmethod
-    def create_with_metadata_builders(
-        backend: type[AttentionBackend],
-        layer_names: list[str],
-        kv_cache_spec: KVCacheSpec,
-        vllm_config: VllmConfig,
-        device: torch.device,
+    def create_metadata_builders(
+        self,
+        vllm_config,
+        device,
+        kernel_block_size: int,
         num_metadata_builders: int = 1,
-    ) -> "AttentionGroup":
-        metadata_builders = [
-            backend.get_builder_cls()(kv_cache_spec, layer_names, vllm_config, device)
+    ):
+        self.metadata_builders = [
+            self.backend.get_builder_cls()(
+                self.kv_cache_spec,
```

Comment:
is it possible to do something like
```
kv_cache_spec = self.kv_cache_spec
if kv_cache_spec.block_size != kernel_block_size:
     kv_cache_spec = kv_cache_spec.with_block_size(kernel_block_size) # creates a new kv_cache_spec with this block_size
```

Reply:
I don't really like having `kv_cache_spec.block_size` be ambiguous (e.g., sometimes being KV cache manager block size, sometimes being kernel block size). Since the kernel block size really depends on the attention backend selection, it should not live in the KVCacheSpec imo. 

If we want to avoid adding `kernel_block_size` to the ctor of every attention backend, we only add a class property to the backends like `requires_kernel_block_size: bool` and then do something like:

```python
builder_extra_args={}
if self.backend.get_builder_cls().requires_kernel_block_size:
    builder_extra_args['kernel_block_size'] = kernel_block_size

self.backend.get_builder_cls()(
    self.kv_cache_spec, 
    self.layer_names, 
    vllm_config, 
    device,
    **builder_extra_args
)
```

Reply:
I don't want people that implement new attention backends to have full context about the difference between `kv_cache_spec.block_size` and `kernel_block_size`. If use your interface, people will still use kv_cache_spec.block_size accidentally.

Reply:
OK, that's a good point. We don't want the attention backend to have to know about this logic. I did another pass at it. Please TAL.